src/sdk/noobaa_s3_client/noobaa_s3_client_sdkv3.js

```diff
@@ -0,0 +1,192 @@
+const cloud_utils = require('../../../src/util/cloud_utils');
+const dbg = require('../../util/debug_module')(__filename);
+const string_utils = require('../../util/string_utils');
+
+// create map of commands
+const map_of_commands = new Map();
+map_of_commands.set('abortMultipartUpload', clients3.AbortMultipartUploadCommand);
+map_of_commands.set('copyObject', clients3.CopyObjectCommand);
+map_of_commands.set('completeMultipartUpload', clients3.CompleteMultipartUploadCommand);
+map_of_commands.set('createMultipartUpload', clients3.CreateMultipartUploadCommand);
+map_of_commands.set('deleteObject', clients3.DeleteObjectCommand);
+map_of_commands.set('deleteObjects', clients3.DeleteObjectsCommand);
+map_of_commands.set('deleteObjectTagging', clients3.DeleteObjectTaggingCommand);
+map_of_commands.set('getObject', clients3.GetObjectCommand);
+map_of_commands.set('getObjectAcl', clients3.GetObjectAclCommand);
+map_of_commands.set('getObjectTagging', clients3.GetObjectTaggingCommand);
+map_of_commands.set('headObject', clients3.HeadObjectCommand);
+map_of_commands.set('listMultipartUploads', clients3.ListMultipartUploadsCommand);
+map_of_commands.set('listObjects', clients3.ListObjectsCommand);
+map_of_commands.set('listObjectsV2', clients3.ListObjectsV2Command);
+map_of_commands.set('listObjectVersions', clients3.ListObjectVersionsCommand);
+map_of_commands.set('listParts', clients3.ListPartsCommand);
+map_of_commands.set('putObject', clients3.PutObjectCommand);
+map_of_commands.set('putObjectAcl', clients3.PutObjectAclCommand);
+map_of_commands.set('putObjectTagging', clients3.PutObjectTaggingCommand);
+map_of_commands.set('uploadPart', clients3.UploadPartCommand);
+map_of_commands.set('uploadPartCopy', clients3.UploadPartCopyCommand);
+
+// we get the op as a string written in Camel case (as as it was in aws sdk v2)
+function get_command(op, params) {
+    const Command = map_of_commands.get(op);
+    // eslint-disable-next-line no-new
+    return new Command(params);
+}
+
+function get_region_from_aws_error(err) {
+    let region;
+
+    if (err.Code === 'AuthorizationHeaderMalformed') {
+        region = err.Region;
+    } else if (err.Code === 'TemporaryRedirect' || err.Code === 'PermanentRedirect') {
+        const endpoint_in_error = err.Endpoint; // structure: {bucket}.s3-{region}.amazonaws.com
+        region = endpoint_in_error.slice(string_utils.index_of_end(endpoint_in_error, '.s3-'), endpoint_in_error.indexOf('.amazonaws.com'));
+    } else {
+        dbg.error('get_region_from_aws_error: Could not get bucket AWS S3 region, got error', err);
+    }
+
+    return region;
+}
+
+/**
+ * S3 Interface, S3 Class and operations in AWS SDK V3:
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3/Interface/S3/
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3/Class/S3/
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/
+ */
+
+class S3ClientSDKV3 {
```

Comment:
why not extend the S3Client and override the send function as we discussed?
This will eliminate the need to redefine all the api methods.
Unless I'm missing something I think it can just be:

```js
/**
 * Same as S3Client but wraps the send method to resubmit the request with the correct region
 * if the server responds that another region should be used.
 */
class S3ClientAutoRegion extends S3Client {

    /**
     * Overriding the send method by wrapping it and fixing the region if the error indicates it.
     * All the methods of the base class call send() to submit an SDK command.
     * @override
     */
    async send(params) {
        try {
           const res = await super.send(params);
        } catch (err) {
            const region = get_region_from_aws_error(err);
            if (!region) throw err;
            this.config.region = region;
            return await super.send(params);
        }
    }

}

```

Reply:
Let's take an example: we have a client that didn't pass the region and his first action is `headObject`.
In HTTP HEAD response body is empty.
Hence in this case we would get the error code 400 without any attached information in the error, I would not have the information to extract on the region and I would throw an error.
This is the way I chose to pay for the first time in `listObjects` with the wrong region and could extract the region from the error.

Anyway, I know of an action that doesn't behave the same (for example, return a different response type) and I would use this class in similar cases.

Reply:
TemporaryRedirect is http 302 or 307 and PermanentRedirect is http 301 or 308 and both send a Location header as a response. Is that not the case for HeadObject with a wrong region?

Reply:
In `headObject` with a wrong region, I get `301: UnknownError` without an error that I can work with.

Reply:
@shirady try this:

```js
/* Copyright (C) 2023 NooBaa */
'use strict';

const { S3 } = require('@aws-sdk/client-s3');

/**
 * Same as S3Client but wraps the send method to resubmit the request with the correct region
 * if the server responds that another region should be used.
 */
class S3ClientAutoRegion extends S3 {

    /**
     * Overriding the send method by wrapping it and fixing the region if the error indicates it.
     * All the methods of the base class call send() to submit an SDK command.
     * @override
     */
    async send(...args) {
        try {
            console.log('\n' + args[0].constructor.name);
            console.log('REGION', this.config.region);
            const res = await super.send(...args);
            console.log('OK', res.$metadata.httpStatusCode);
            return res;
        } catch (err) {
            console.log('ERROR', err.message);
            const region = err.$response?.headers?.['x-amz-bucket-region'];
            if (!region) throw err;
            console.log('NEW REGION', region);
            this.config.region = region;
            const res = await super.send(...args);
            console.log('NEW OK', res.$metadata.httpStatusCode);
            return res;
        }
    }
}

async function main() {
    const s3 = new S3ClientAutoRegion();
    const Bucket = 'guymguym';
    const Key = 'lala';
    const wrong_region = 'eu-west-2';

    s3.config.region = wrong_region;
    await s3.listBuckets({});

    s3.config.region = wrong_region;
    await s3.headBucket({ Bucket });

    s3.config.region = wrong_region;
    await s3.listObjects({ Bucket });

    s3.config.region = wrong_region;
    await s3.putBucketTagging({ Bucket, Tagging: { TagSet: [] } });

    s3.config.region = wrong_region;
    await s3.headObject({ Bucket, Key });

    s3.config.region = wrong_region;
    await s3.getObject({ Bucket, Key });

    s3.config.region = wrong_region;
    await s3.putObjectTagging({ Bucket, Key, Tagging: { TagSet: [] } });

    s3.config.region = wrong_region;
    await s3.putObject({ Bucket, Key, Body: Key });

    s3.destroy();
}

main();

```

Reply:
@guymguym Thank you
I have a couple of concerns:
1) I don't know in what action I wouldn't get the region in the response header - and it will be a bug... (very specific: create a client without region first using an action that in its response doesn't have 'x-amz-bucket-region').
2) Using it I would mean losing the control that I had when I explicitly defined for every action what to run and used a flag to check if I already updated the region (I defined the actions under the class).
3) I saw a code snippet in Python ([here](https://www.programcreek.com/python/?CodeExample=get%20bucket%20region)) that does two extra checks: checks if the region appears in the error and if both of them not runs a `headBucket` and take it from its header (align to what I did at first). The last check cannot be used in this class as is now because it works for every send (and not a specific use to get the region).

Reply:
> I don't know in what action I wouldn't get the region in the response header - and it will be a bug... (very specific: create a client without region first using an action that in its response doesn't have 'x-amz-bucket-region').

I prefer this optimistic approach, because the other way would be that every operation would have to start by calling GetBucketLocation API to find the bucket location first (but definitely shouldn't call ListObjects as you have currently in the PR which is way to heavy). Also calling any other call than then intended API is prone to Bucket Policy errors since in AWS every IAM can be allowed to perform different API calls. The safest approach is to invoke only the requested operation and parse the response headers. You can see that I tested this works for a handful of operations above, and all send this header as expected. If AWS doesn't reply that header for a specific API - we can workaround it.

> Using it I would mean losing the control that I had when I explicitly defined for every action what to run and used a flag to check if I already updated the region (I defined the actions under the class).

I don't want us to have to maintain the list of S3 commands - I don't think this adds any real control. This handling can and should be dynamic. If you want to assert any behavior - we can add a test.

> I saw a code snippet in Python ([here](https://www.programcreek.com/python/?CodeExample=get%20bucket%20region)) that does two extra checks: checks if the region appears in the error and if both of them not runs a headBucket and take it from its header (align to what I did at first). The last check cannot be used in this class as is now because it works for every send (and not a specific use to get the region).

I would start just with parsing the header. If we see that calling GetBucketLocation or HeadBucket is needed we can always add that inside of our send() call as a fallback option to get the region. I don't see a problem to add such calls there as needed.